packages/strapi-plugin-graphql/services/Loaders.js

```diff
@@ -0,0 +1,187 @@
+    Object.keys(strapi.models)
+      .filter(model => model !== 'core_store')
+      .forEach(model => {
+        (strapi.models[model].associations || []).forEach(association => this.createLoader(association.collection || association.model, association.plugin));
+      });
+
+    // Reproduce the same pattern for each plugin.
+    Object.keys(strapi.plugins).forEach(plugin => {
+      Object.keys(strapi.plugins[plugin].models).forEach(model => {
+        (strapi.plugins[plugin].models[model].associations || []).forEach(association => this.createLoader(association.collection || association.model, association.plugin));
+      });
+    });
+  },
+
+  createLoader: function(model, plugin) {
+    const name = plugin ? `${plugin}__${model}`: model;
+
+    this.loaders[name] = new DataLoader(keys => {
+      return new Promise(async (resolve, reject) => {
+        try {
+          // Extract queries from keys and merge similar queries.
+          const { queries, map } = this.extractQueries(model, _.cloneDeep(keys));
+          // Run queries in parallel.
+          const results = await Promise.all(queries.map((query) => this.makeQuery(model, query)));
+          // Use to match initial queries order.
+          const data = this.mapData(model, keys, map, results);
+
+          resolve(data);
+        } catch (e) {
+          reject(e);
+        }
+      });
+    }, {
+      cacheKeyFn: (key) => {
+        return _.isObjectLike(key) ? JSON.stringify(_.cloneDeep(key)) : key;
+      }
+    });
+  },
+
+  mapData: function(model, originalMap, map, results) {
+    // Use map to re-dispatch data correctly based on initial keys.
+    return originalMap.map((query, index) => {
+      // Find the index of where we should extract the results.
+      const indexResults = map.findIndex(queryMap => queryMap.indexOf(index) !== -1);
+      const data = results[indexResults];
+
+      // Retrieving referring model.
+      const ref = this.retrieveModel(model, query.options.source);
+
+      if (query.single) {
+        // Return object instead of array for one-to-many relationship.
+        return data.find(entry => entry[ref.primaryKey].toString() === (query.params[ref.primaryKey] || '').toString());
+      }
+
+      // Extracting ids from original request to map with query results.
+      const ids = (() => {
+        if ( _.get(query.options, `query.${ref.primaryKey}`)) {
+          return  _.get(query.options, `query.${ref.primaryKey}`);
+        }
+
+        // Single object to retrieve (one-to-many).
+        const alias = _.first(Object.keys(query.options.query));
+
+        return {
+          alias,
+          value: _.get(query.options, `query.${alias}`)
+        };
+      })();
+      
+      if (!_.isArray(ids)) {
+        return data
+          .filter(entry => entry[ids.alias].toString() === ids.value.toString())
+          .slice((query.options.start || query.options.skip), (query.options.start || query.options.skip) + query.options.limit);
```

Comment:
Don't try to support two way of doing things it'll just create problems afterward, if there is no way of doing so, I'd recommend to create a utility function that does it for you or simply create a constant where you do this OR block once:
```JS
const skip = query.options.start || query.options.skip;
```

Reply:
I agree. I've created a constant as you recommend, we shouldn't support both ways to do this thing in the future.